Add basic dimming functionality

farwyler opened the pull request:
This adds a first iteration of color dimming to the editor.

You can:
1. Dim the colors of all unfocused views
2. Dim the colors in the background of Overlays (pickers, etc.)

The following dimming functions are implemented:
- "dim" just set the modifier flag "dim" of all cells
- "lighten" / "darken" actually change the shade of the fg/bg colors. this only works for themes applying RGB colors, not ANSI colors.

Config example:

### `[editor.dim]` Section

Enable dimming in certain areas. Disabled by default. Enable by assigning:
- 0 = set text modifier flag DIM
- negative value = darken colors (-127 = 100% darker)
- positive value = lighten colors (+127 = 100% lighter)

> darken/lighten colors only works with themes that set rgb colors, not indexed colors.

| Key | Description | Default |
|--|--|---------|
| `overlay-backdrops` | Enable dimming and set shade of content behind overlays. | `None` |
| `unfocused-views`| Enable dimming and set shade of unfocused editor views. | `None` |

```toml
[editor.dim]
# darken backdrop of overlays
overlay-backdrops = -25
# dim fg color of unfocused views
unfocused-views = 0
```


Screenshots:

![Screenshot 2022-06-11 at 19 15 28](https://user-images.githubusercontent.com/1705805/173198014-bb7535fe-1616-4aff-bf86-1afab8e10227.png)
![Screenshot 2022-06-11 at 19 16 09](https://user-images.githubusercontent.com/1705805/173198015-1380663d-cccf-4bdc-93b1-935c2e8bd98e.png)

Todo:
* [x] feedback
* [x] documentation
* [x] some optimizations


Looking for feedback.
Do we actually want this functionality in the tui?
Leave this to alternative frontends or plugins later?
Could this be used with  #1187 ?

sudormrfbin commented:
I guess this is a nice to have, but a config for it seems overkill, especially the kind of dimming functions to use. If this does get merged, we would also want to avoid pulling in a dependency (`palette`) simply to dim colors.

---

It does feel like a good usability improvement after trying it out though, so I'm more or less on the fence now 😛

farwyler commented:
thanks @CptPotato  and @sudormrfbin for the feedback.
i have simplified the implementation and removed the dependency.
new configuration looks like [this](https://github.com/farwyler/helix/blob/dimming/book/src/configuration.md#editordim-section).

yvt commented on helix-tui/src/buffer.rs:
```diff
@@ -583,6 +583,50 @@ impl Buffer {
+    /// shade < 0 -> darken rgb color
+    /// shade > 0 -> lighten rgb color
+    pub fn dim(&mut self, area: Rect, shade: i8) {
+        let shade_amount = shade.unsigned_abs().saturating_mul(2);
+        let shade_fn = if shade < 0 {
+            u8::saturating_sub
+        } else {
+            u8::saturating_add
+        };
+        // TODO: replace with saturating_add_signed when mixed_integer_ops (https://github.com/rust-lang/rust/issues/87840) get stabilized.
+        let shaded = |v| shade_fn(v, shade_amount);
```
I'd like to note that saturating addition/subtraction has a very different effect than that of [`palette::Shade::{lighten, darken}`](https://docs.rs/palette/0.6.0/src/palette/rgb/rgb.rs.html#535-580), which was used before 9ad01a06232ae03ed99cbd601764aab29fe5b379, and is more likely to cause undesirable effects such as clipping, saturation boosting, and hue shifting (examples shown below).

- ![](https://img.shields.io/badge/-%E3%80%80-64460a?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-0a6400?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-325014?style=for-the-badge) `[(100, 70, 10), (10, 100, 0), (50, 80, 20)]` (various dark tones) → ![](https://img.shields.io/badge/-%E3%80%80-000?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-000?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-000?style=for-the-badge) `[(0, 0, 0); 3]` (black)
- ![](https://img.shields.io/badge/-%E3%80%80-c86464?style=for-the-badge) `(200, 100, 100)` (desaturated red) → ![](https://img.shields.io/badge/-%E3%80%80-640000?style=for-the-badge) `(100, 0, 0)` (darker, but more vivid red)
- ![](https://img.shields.io/badge/-%E3%80%80-c86400?style=for-the-badge) `(200, 100, 0)` (orange) → ![](https://img.shields.io/badge/-%E3%80%80-640000?style=for-the-badge) `(100, 0, 0)` (dark red)

Here's a possible implementation based on alpha compositing without gamma correction (warning: untested):

```suggestion
        let alpha = i32::from(shade).unsigned_abs();
        let (src_factor, dst_factor) = (alpha, 256 - alpha);
        let src = if shade > 0 { 255u32 } else { 0u32 } * src_factor;
        let shaded = |dst_color: u8| ((u32::from(dst_color) * dst_factor + src) >> 8) as u8;
```

... which should produce colors like these:

- ![](https://img.shields.io/badge/-%E3%80%80-64460a?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-0a6400?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-325014?style=for-the-badge) `[(100, 70, 10), (10, 100, 0), (50, 80, 20)]` (various dark tones) → ![](https://img.shields.io/badge/-%E3%80%80-322305?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-053200?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-19280a?style=for-the-badge) `[(50, 35, 5), (5, 50, 0), (25, 40, 10)]` (various darker tones)
- ![](https://img.shields.io/badge/-%E3%80%80-c86464?style=for-the-badge) `(200, 100, 100)` (desaturated red) → ![](https://img.shields.io/badge/-%E3%80%80-643232?style=for-the-badge) `(100, 50, 50)` (darker, still desaturated red)
- ![](https://img.shields.io/badge/-%E3%80%80-c86400?style=for-the-badge) `(200, 100, 0)` (orange) → ![](https://img.shields.io/badge/-%E3%80%80-643200?style=for-the-badge) `(100, 50, 0)` (dark orange)

farwyler replied:
This works great. Thank you 🙏

yvt commented on helix-tui/src/buffer.rs:
```diff
@@ -583,6 +583,50 @@ impl Buffer {
+        // TODO: replace with saturating_add_signed when mixed_integer_ops (https://github.com/rust-lang/rust/issues/87840) get stabilized.
+        let shaded = |v| shade_fn(v, shade_amount);
+
+        let mut cache: HashMap<Color, Color> = HashMap::new();
+        let mut modify = |col| {
+            let result = cache.entry(col).or_insert_with(|| {
```
I suspect that the color transformation is faster than SipHash 2-4 hashing + memory allocation.

farwyler replied:
i just benched it. you are right. way too much overhead at these tiny resolutions.
thanks for catching this. removed it in 8625892af12c3f92e4f6035835169ace0f95e44c

yvt commented on helix-tui/src/buffer.rs:
```diff
@@ -583,6 +582,35 @@ impl Buffer {
+
+    /// Apply shade to all cells in area.
+    ///
+    /// shade = 0 -> set text modifier DIM
+    /// shade < 0 -> darken rgb color
+    /// shade > 0 -> lighten rgb color
```
Regular line breaks in doc comments might not render as line breaks (CommonMark [§6.8](https://spec.commonmark.org/0.30/#hard-line-breaks))

```suggestion
    /// - `shade = 0`: set text modifier `DIM`
    /// - `shade < 0`: darken rgb color
    /// - `shade > 0`: lighten rgb color
```

archseer commented:
Doing dimming across an entire area after the fact seems a bit of a hack -- I'd rather calculate the dimming inside render_view, similar to how we treat active/unfocused views by passing a bool into that function.

farwyler commented:
@archseer You are right, the post processing is quite hacky. I wanted to keep the dimming out of the rendering functions if possible, as it introduces code complexity there for something that is a graphical gimmick, at best a slight UX improvement.

With bb4fec7ee8ecaf596f932bbc18806ab1439210a7 i've tried another an implementation that does the dimming ad-hoc, with just slight adjustments to the main render loop and a little state in the drawing buffer.

pascalkuthe commented:
clsoing this one as stale. Almost the entire rendering code has been removed since so this would need to start over from scratch. Thank you for contributing!
